src/integrationTests/clientSideTokenGeneration.test.ts

```diff
@@ -0,0 +1,188 @@
+  serverKeyPair = await crypto.subtle.generateKey(
+    {
+      name: 'ECDH',
+      namedCurve: NAME_CURVE,
+    },
+    false,
+    ['deriveKey']
+  );
+
+  const serverPubKeyArray = await crypto.subtle.exportKey('spki', serverKeyPair.publicKey);
+  serverPublicKey = `UID2-X-L-` + bytesToBase64(new Uint8Array(serverPubKeyArray));
+});
+
+beforeEach(() => {
+  xhrMock = new mocks.XhrMock(sdkWindow);
+  mocks.setCookieMock(sdkWindow.document);
+  removeUid2LocalStorage();
+});
+
+afterEach(() => {
+  mocks.resetFakeTime();
+});
+
+const removeUid2LocalStorage = mocks.removeUid2LocalStorage;
+
+describe('Client-side token generation Tests', () => {
+  const scenarios = [
+    {
+      name: 'setIdentityFromEmail',
+      setInvalidIdentity: () => uid2.setIdentityFromEmail('test.com', mocks.makeCstgOption()),
+      setIdentity: (serverPublicKey?: string) =>
+        uid2.setIdentityFromEmail('test@example.com', mocks.makeCstgOption({ serverPublicKey })),
+    },
+    {
+      name: 'setIdentityFromEmailHash',
+      setInvalidIdentity: () =>
+        uid2.setIdentityFromEmailHash('test@example.com', mocks.makeCstgOption()),
+      setIdentity: (serverPublicKey?: string) =>
+        uid2.setIdentityFromEmailHash(
+          'lz3+Rj7IV4X1+Vr1ujkG7tstkxwk5pgkqJ6mXbpOgTs=',
+          mocks.makeCstgOption({ serverPublicKey })
+        ),
+    },
+    {
+      name: 'setIdentityFromPhone',
+      setInvalidIdentity: () => uid2.setIdentityFromPhone('12345678910', mocks.makeCstgOption()),
+      setIdentity: (serverPublicKey?: string) =>
+        uid2.setIdentityFromPhone('+12345678910', mocks.makeCstgOption({ serverPublicKey })),
+    },
+    {
+      name: 'setIdentityFromPhoneHash',
+      setInvalidIdentity: () =>
+        uid2.setIdentityFromPhoneHash('+12345678910', mocks.makeCstgOption()),
+      setIdentity: (serverPublicKey?: string) =>
+        uid2.setIdentityFromPhoneHash(
+          'kVJ+4ilhrqm3HZDDnCQy4niZknvCoM4MkoVzZrQSdJw=',
+          mocks.makeCstgOption({ serverPublicKey })
+        ),
+    },
+  ];
+
+  scenarios.forEach((scenario) => {
+    describe(scenario.name, () => {
+      beforeEach(() => {
+        uid2 = new UID2();
+        uid2.init({});
+      });
+
+      describe('When invalid identity is provided', () => {
+        test('should throw error', (done) => {
+          uid2.callbacks.push(async (eventType: EventType) => {
+            if (eventType === EventType.InitCompleted) {
+              try {
+                await scenario.setInvalidIdentity();
+              } catch (err) {
+                done();
+              }
+            }
+          });
+        });
+      });
+
+      describe('When valid identity is provided', () => {
+        const setIdentityInCallback = () => {
+          let setIdentityResultPromise;
+          const callback = async (eventType: EventType) => {
+            if (eventType === EventType.InitCompleted) {
+              setIdentityResultPromise = scenario.setIdentity(serverPublicKey);
```

Comment:
Do you mean when `setIdentityInCallback()` is called, but setIdentityResultPromise is still since the callback hasn't been triggered? 
I don't think this is possible, since `uid2.init({})` is not async, so the initialized should complete in the beforeEach hook. I think using a callback to wrap the setIdentity functions makes this confusing. I removed callbacks from the tests